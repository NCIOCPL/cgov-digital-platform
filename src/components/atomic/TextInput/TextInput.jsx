import React from 'react';
import PropTypes from 'prop-types';
import Utilities from '../../../utilities/utilities';
import InputLabel from '../InputLabel';
import './TextInput.scss';

class TextInput extends React.Component {
  static propTypes = {
    action: PropTypes.func,
    allowedChars: PropTypes.object,
    classes: PropTypes.string,
    enableSpellCheck: PropTypes.bool,
    errorMessage: PropTypes.string,
    id: PropTypes.string.isRequired,
    inputHelpText: PropTypes.string,
    isValid: PropTypes.bool,
    label: PropTypes.string.isRequired,
    labelHidden: PropTypes.bool,
    labelHint: PropTypes.string,
    maxLength: PropTypes.number,
    placeHolder: PropTypes.string,
    required: PropTypes.bool,
    type: PropTypes.oneOf([
      'text',
      'email',
      'password',
      'search',
      'url',
      'date',
      'month',
      'tel',
      'week',
      'number',
    ]),
    value: PropTypes.string,
    validators: PropTypes.array
  };

  static defaultProps = {
    action: () => {},
    classes: '',
    type: 'text',
    enableSpellCheck: false,
    required: false,
  };

  constructor(props) {
    super(props);

    let pristine = true;
    if (this.props.value || this.props.errorMessage) {
      pristine = false;
    }

    this.state = {
      value: this.props.value || '',
      isPristine: pristine,
      isValid: this.props.isValid ? true : false,
      hasError: this.props.errorMessage ? true : false,
      errorMessage: this.props.errorMessage,
    };

    // Generate an HTML ID if one was not provided
    this.id = this.props.id || Utilities.uniqueIdForComponent();
  }

  componentDidUpdate(prevProps) {
    // If a errorMessage is passed after initial render, adjust the state accordingly
    if (prevProps.errorMessage !== this.props.errorMessage) {
      this.setState({
        isPristine: false,
        isValid: false,
        hasError: true,
        errorMessageBody: this.props.errorMessage,
      });
    }
  }

  render() {
    let error,
      helpText,
      ariaLabel = null;
    if (this.state.hasError) {
      error = (
        <span className="cts-input__error" role="alert">
          {this.state.errorMessage}
        </span>
      );
    }
    if (this.props.inputHelpText) {
      helpText = (
        <span className="cts-input__help-text">{this.props.inputHelpText}</span>
      );
    }

    ariaLabel = this.props.labelHidden
      ? { 'aria-label': this.props.label }
      : { 'aria-labelledby': this.props.id + '-label' };

    return (
      <>
        {this.props.labelHidden ? null : (
          <InputLabel
            label={this.props.label}
            labelHint={this.props.labelHint}
            htmlFor={this.id}
            hasError={this.state.hasError}
            required={this.props.required}
          />
        )}
        <input
          id={this.id}
          type={this.props.type}
          value={this.state.value}
          className={`cts-input ${this.props.classes} ${
            this.state.isValid ? 'cts-input--success' : ''
          }`}
          required={this.props.required}
          maxLength={this.props.maxLength}
          placeholder={this.props.placeHolder}
          aria-required={this.props.required}
          onBlur={this._handleBlur.bind(this)}
          onChange={this._handleChange.bind(this)}
          spellCheck={this.props.enableSpellCheck ? true : false}
          {...ariaLabel}
        />
        {error}
        {helpText}
      </>
    );
  }

  _validate() {
    let validators = this.props.validators;
    // Check if field empty
    if (!this.state.value) {
      // If it's required, say so
      if (this.props.required) {
        this.setState({
          hasError: true,
          isValid: false,
          errorMessage: 'This field is required',
        });
      } else {
        // is empty so reset isValid and hasError
        this.setState({
          isValid: false,
          hasError: false,
          errorMessage: null,
        });
      }
    }
    // If validator(s) were sent as a prop, test them next
    else if (validators) {
      // eslint-disable-next-line
      for (let validator of validators) {
        // check is validator is forced 'no validation'
        if (!validator.pattern && !validator.isValid(this.state.value)) {
          this.setState({
            hasError: true,
            isValid: false,
            errorMessage: validator.message,
          });
          break;
        } else {
          this.setState({
            hasError: false,
            isValid: true,
          });
        }
      }
    }
    // must be required field with a value, so no error
    else {
      this.setState({
        hasError: false,
        errorMessage: null,
      });
    }
  }

  //  onBlur event on input
  _handleBlur() {
    if (
      (this.props.required || this.props.validators) &&
      !this.state.isPristine
    ) {
      this._validate();
    }
  }

  //  his function runs every time the user changes the contents of the input.
  //  @param {event} event The event
  _handleChange(event) {
    // Check if allowedChars validator exists. If it does, check the last char
    // entered against the validator. If validation fails, return thereby preventing
    // the value from being added to the state.
    if (this.props.allowedChars) {
      let input = event.target.value.slice(-1);
      if (!this.props.allowedChars.isValid(input)) {
        return;
      }
    }

    // Call action handler prop
    this.props.action(event.target.value);

    // Commit the input's value to state.value.
    this.setState({ value: event.target.value }, () => {
      // React docs suggest this callback should generally go in ComponentDidUpdate,
      // however since both this callback actions update the state, they must
      // go here because changing state in ComponentDidUpdate would cause a
      // recursive loop and blow up the call stack
      if (this.state.value && this.state.isPristine) {
        this.setState({ isPristine: false });
      }
      // if
      if (this.state.hasError || this.state.isValid) {
        this._validate();
      }
    });
  }
}

export default TextInput;
