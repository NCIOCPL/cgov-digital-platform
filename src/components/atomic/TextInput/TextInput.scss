.cts-input {
  &[type='date'],
  &[type='datetime-local'],
  &[type='datetime'],
  &[type='email'],
  &[type='month'],
  &[type='number'],
  &[type='password'],
  &[type='search'],
  &[type='tel'],
  &[type='text'],
  &[type='time'],
  &[type='url'],
  &[type='week'] {
    background-color: white;
    margin: 0;
    &:focus {
      background-color: white;
    }
  }

  &__help-text {
    display: inline-block;
    margin-top: 5px;
    font-size: 13px;
    line-height: 1.4;
  }

  &__error-message {
    padding-bottom: 5px;
    color: #b50909;
    display: block;
    font-weight: 700;
    line-height: 1.4;
    font-size: 14px;
  }

  &.cts-input--modified {
    background: rgba(254, 230, 133, .4);
  }

  &.cts-input--error {
    border: 2px solid #b50909;
  }

  .faux-select {
    background-image: url('data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9IjAgMCAxMCAxNCI+PGcgZmlsbD0iIzNkNDU1MSIgZmlsbC1ydWxlPSJldmVub2RkIj48cGF0aCBkPSJNLjYyNjM5NjI4IDUuOTk5OTM4OTZIOS4zNzI3Mzg5Yy4yNTMzOTE3My4wMDEzMDg1Ni40ODI0MTYtLjE1MDc0MjY0LjU3OTU2MjAyLS4zODQ3NzYyNC4wOTcxNDU5OC0uMjM0MDMzNi4wNDMxMjI5NC0uNTAzNTc2MTYtLjEzNjcwOTIyLS42ODIwOTY0TDUuNDQwMTgxNS4xODM1NDM1QzUuMzI0NzU2MDYuMDY2MTMxMTYgNS4xNjcwMDgzIDAgNS4wMDIzNjExNCAwYy0uMTY0NjQ3MTYgMC0uMzIyMzk0OS4wNjYxMzExNy0uNDM3ODIwMzcuMTgzNTQzNUwuMTgzNTQzNSA0LjkzMzA2NjMyQy4wNjYxMzExNiA1LjA0ODQ5MTggMCA1LjIwNjIzOTUzIDAgNS4zNzA4ODY3cy4wNjYxMzExNy4zMjIzOTQ5Mi4xODM1NDM1LjQzNzgyMDM4Yy4xMTQ2NDQ0LjEyMjQ5MzQuMjc1MDgwMDYuMTkxNzcyNDQuNDQyODUyNzguMTkxMjMxODh6TTkuMzcyNzM4OSA3Ljk5OTk2ODVILjYyNjM5NjNjLS4xNjYyNjUzLS4wMDMxNjE4My0uMzI2NDc2MDguMDYyMzc4OTQtLjQ0Mjg1MjguMTgxMTY3MDVDLjA2NjEzMTIgOC4yOTY1NjEgMCA4LjQ1NDMwODc1IDAgOC42MTg5NTU5MnMuMDY2MTMxMTcuMzIyMzk0OTIuMTgzNTQzNS40Mzc4MjAzN2w0LjM4MDk5NzI2IDQuNzU5NjgwMkM0LjY3OTk2NjIyIDEzLjkzMzg2ODggNC44Mzc3MTM5NiAxNCA1LjAwMjM2MTEzIDE0cy4zMjIzOTQ5Mi0uMDY2MTMxMi40Mzc4MjAzOC0uMTgzNTQzNWw0LjM3NTQxMDItNC43NDk2MTUzN2MuMTc5ODMyMTYtLjE3ODUyMDIzLjIzMzg1NTItLjQ0ODA2Mjc4LjEzNjcwOTIyLS42ODIwOTYzOC0uMDk3MTQ2MDItLjIzNDAzMzYtLjMyNjE3MDMtLjM4NjA4NDgtLjU3OTU2Mi0uMzg0Nzc2MjV6Ii8+PC9nPjwvc3ZnPg=='),
      linear-gradient(transparent, transparent);
    background-repeat: no-repeat;
    appearance: none;
    background-color: white;
    background-position: right 0.75rem center;
    background-size: 0.5rem;
    padding-right: 2rem;
  }
}

.cts-input-group {
  + .cts-input-group {
    margin-top: 1rem;
  }
}
