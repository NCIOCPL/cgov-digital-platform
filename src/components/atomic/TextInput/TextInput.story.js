import generateStories from '../../../../.storybook/generateStories';
import TextInput from './TextInput';
import { text } from '@storybook/addon-knobs';

generateStories({
  component: { TextInput },
  storyName: 'Elements | TextInput',
  defaultProps: {
    id: 'mockid',
    label: 'Input Label',
    type: "text",
    placeHolder: 'placeholder text',
    required: false,
    labelHidden: false
  },
  stories: [
    {
      name: 'default'
    }
  ]
});
