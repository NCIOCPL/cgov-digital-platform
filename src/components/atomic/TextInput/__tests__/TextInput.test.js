import React from 'react';
import { mount } from 'enzyme';
import TextInput from '../TextInput.jsx';

jest.unmock('../TextInput.jsx');

describe('TextInput', () => {
  let wrapper = null;
  const label = 'label goes here';
  const testId = 'myId';

  beforeEach(function() {
    wrapper = mount(<TextInput id={testId} label={label} />);
  });

  it('renders without error', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('is defined', () => {
    expect(wrapper).toBeDefined();
  });

  it('renders the label text correctly', () => {
    const labelElement = wrapper.find('label');
    expect(labelElement.text()).toBe(label);
  });

  it('has an id attribute', () => {
    expect(typeof wrapper.find('input').props().id).toBe('string');
  });

  it('has a matching id and htmlFor attributes', () => {
    const id = wrapper.find('input').props().id;
    const labelHtmlFor = wrapper.find('label').props().htmlFor;
    expect(labelHtmlFor).toBe(id);
  });

  it('accepts a value property', () => {
    const testValue = 'testing';
    wrapper = mount(<TextInput id={testId} value={testValue} label={label} />);
    const value = wrapper.find('input').props().value;
    expect(value).toBe(testValue);
  });

  it('accepts the required property', () => {
    wrapper = mount(<TextInput required id={testId} label={label} />);
    expect(wrapper.find('label').hasClass('cts-label--required')).toBe(true);
  });
});
