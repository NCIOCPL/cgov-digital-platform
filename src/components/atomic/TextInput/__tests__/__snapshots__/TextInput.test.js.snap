// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput renders without error 1`] = `
<TextInput
  action={[Function]}
  classes=""
  disabled={false}
  enableSpellCheck={false}
  id="myId"
  label="label goes here"
  required={false}
  type="text"
>
  <div
    className="cts-input-group "
  >
    <InputLabel
      hasError={false}
      htmlFor="myId"
      label="label goes here"
      required={false}
    >
      <label
        className="cts-label"
        htmlFor="myId"
        id="myId-label"
      >
        label goes here
      </label>
    </InputLabel>
    <input
      aria-labelledby="myId-label"
      aria-required={false}
      className="cts-input  "
      disabled={false}
      id="myId"
      onBlur={[Function]}
      onChange={[Function]}
      required={false}
      spellCheck={false}
      type="text"
      value=""
    />
  </div>
</TextInput>
`;
